File: mototrack-server.py
#!/usr/bin/env python3
"""MotoTrack Server v9 - Threaded with proper chunk serving"""
import http.server
import socketserver
import os
import json
import threading

PORT = 3000
BASE = os.path.dirname(os.path.abspath(__file__))

# Load HTML
try:
    with open(os.path.join(BASE, '.next', 'cached-index.html'), 'rb') as f:
        INDEX_HTML = f.read()
except:
    INDEX_HTML = b'<html><body><h1>MotoTrack</h1><p>Building...</p></body></html>'

MIME = {
    '.js': 'application/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.woff2': 'font/woff2',
    '.woff': 'font/woff',
    '.webp': 'image/webp',
    '.wasm': 'application/wasm',
    '.map': 'application/json',
    '.txt': 'text/plain',
    '.html': 'text/html',
}

_cache = {}
_cache_lock = threading.Lock()

def load_file(path):
    with _cache_lock:
        if path in _cache:
            return _cache[path]
    try:
        if not os.path.isfile(path):
            return None
        with open(path, 'rb') as f:
            data = f.read()
        ext = os.path.splitext(path)[1].lower()
        ct = MIME.get(ext, 'application/octet-stream')
        with _cache_lock:
            _cache[path] = (data, ct)
        return (data, ct)
    except:
        return None

def preload():
    count = 0
    # Pre-load all static files
    for root, dirs, files in os.walk(os.path.join(BASE, '.next', 'static')):
        for f in files:
            fp = os.path.join(root, f)
            if os.path.getsize(fp) < 5_000_000:  # Skip files > 5MB
                load_file(fp)
                count += 1
    for root, dirs, files in os.walk(os.path.join(BASE, 'public')):
        for f in files:
            fp = os.path.join(root, f)
            if os.path.getsize(fp) < 5_000_000:
                load_file(fp)
                count += 1
    return count

MOCK_USER = {
    'id': 'demo1',
    'name': 'Miran M.',
    'email': 'miran@rever.si',
    'avatar': None,
    'bike': 'Yamaha MT-07',
    'bio': 'Motociklisticni navdusenec'
}

API_GET = {
    '/api/init': lambda: {'users': [MOCK_USER], 'rides': [], 'routes': [], 'defaultUser': MOCK_USER, 'needsSeed': False, 'leaderboard': [{'id': 'demo1', 'name': 'Miran M.', 'totalDistance': 0, 'totalRides': 0}]},
    '/api/notifications': lambda: [],
    '/api/achievements': lambda: {'earned': [], 'newlyEarned': []},
    '/api/sos': lambda: {'ok': True},
    '/api/seed': lambda: {'seeded': False},
    '/api/leaderboard': lambda: [{'id': 'demo1', 'name': 'Miran M.', 'totalDistance': 0, 'totalRides': 0}],
    '/api/rides': lambda: [],
    '/api/routes': lambda: [],
    '/api/comments': lambda: [],
    '/api/settings': lambda: {'unitSystem': 'metric', 'autoPauseEnabled': True, 'autoPauseSpeedThreshold': 5, 'wakelockEnabled': True, 'hideStartEnd': False},
    '/api/stats': lambda: {'totalRides': 0, 'totalDistance': 0, 'totalDuration': 0, 'avgSpeed': 0, 'maxSpeed': 0},
    '/api/weather': lambda: None,
    '/api/fuel': lambda: [],
    '/api/fuel-prices': lambda: [],
    '/api/balkan-roads': lambda: [],
    '/api/events': lambda: [],
    '/api/challenges': lambda: [],
    '/api/favorites': lambda: [],
    '/api/friends': lambda: [],
    '/api/ride-score': lambda: {'score': 0, 'breakdown': {}},
    '/api/curvy-roads': lambda: [],
    '/api/map-styles': lambda: [],
    '/api/camps': lambda: [],
    '/api/expenses': lambda: [],
    '/api/videos': lambda: [],
    '/api/subscription': lambda: {'plan': 'free'},
}

class Handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        path = self.path.split('?')[0]
        
        if path in ('/', ''):
            self._send(200, INDEX_HTML, 'text/html; charset=utf-8', 60)
        elif path.startswith('/_next/static/'):
            # Map /_next/static/... to .next/static/...
            fp = os.path.join(BASE, '.next', 'static', path.replace('/_next/static/', '', 1))
            r = load_file(fp)
            if r:
                self._send(200, r[0], r[1], 31536000)
            else:
                self._send(404, b'Not Found', 'text/plain', 0)
        elif path.startswith('/_next/'):
            # Other _next resources (like _next/data)
            fp = os.path.join(BASE, path.lstrip('/'))
            r = load_file(fp)
            if r:
                self._send(200, r[0], r[1], 3600)
            else:
                # Fall back to index for SPA routing
                self._send(200, INDEX_HTML, 'text/html; charset=utf-8', 60)
        elif path.startswith('/api/'):
            h = API_GET.get(path)
            d = h() if h else None
            self._send(200, json.dumps({'data': d}).encode(), 'application/json', 0)
        elif path == '/sw.js':
            r = load_file(os.path.join(BASE, 'public', 'sw.js'))
            if r:
                self._send(200, r[0], 'application/javascript', 0)
            else:
                self._send(200, b'self.addEventListener("fetch",function(){});', 'application/javascript', 0)
        else:
            # Try public folder first, then fall back to index.html
            fp = os.path.join(BASE, 'public', path.lstrip('/'))
            r = load_file(fp)
            if r:
                self._send(200, r[0], r[1], 3600)
            else:
                self._send(200, INDEX_HTML, 'text/html; charset=utf-8', 60)
    
    def do_POST(self):
        path = self.path.split('?')[0]
        if path.startswith('/api/'):
            h = API_GET.get(path)
            d = h() if h else None
            self._send(200, json.dumps({'data': d}).encode(), 'application/json', 0)
        else:
            self._send(200, b'{"data":null}', 'application/json', 0)
    
    def do_PUT(self):
        self.do_POST()
    
    def do_DELETE(self):
        self.do_POST()
    
    def do_PATCH(self):
        self.do_POST()
    
    def _send(self, code, data, ct, max_age):
        try:
            self.send_response(code)
            self.send_header('Content-Type', ct)
            self.send_header('Content-Length', str(len(data)))
            self.send_header('Cache-Control', f'public, max-age={max_age}')
            self.send_header('Connection', 'close')
            self.end_headers()
            self.wfile.write(data)
        except:
            pass
    
    def log_message(self, format, *args):
        pass  # Suppress logging

class ThreadedServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    allow_reuse_address = True
    request_queue_size = 10
    daemon_threads = True

if __name__ == '__main__':
    count = preload()
    print(f'> MotoTrack v9 on :{PORT} | {count} files cached | threaded')
    server = ThreadedServer(('', PORT), Handler)
    server.serve_forever()

File: mototrack-server.js
#!/usr/bin/env node
/**
 * MotoTrack Ultra-Lightweight Server for Sandbox
 * 
 * Strategy: NO Prisma, NO Next.js runtime. Just static file serving + API stubs.
 * The client-side JavaScript handles empty data with loading states.
 * This keeps RSS at ~50MB which the sandbox allows.
 * 
 * For full API functionality, use the Vercel deployment.
 */

var http = require('http');
var fs = require('fs');
var path = require('path');

var PORT = parseInt(process.env.PORT || '3000', 10);
process.on('uncaughtException', function(e) { console.error('[ERR]', e.message); });
process.on('unhandledRejection', function(r) { console.error('[REJECT]', String(r)); });

// Load cached HTML
var cachedIndex = null;
try {
  cachedIndex = fs.readFileSync(path.join(__dirname, '.next', 'cached-index.html'));
  console.log('[STATIC] Index: ' + cachedIndex.length + ' bytes');
} catch (e) {
  cachedIndex = Buffer.from('<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>MotoTrack</title></head><body><div id="__next"></div></body></html>');
}

var MIME = {
  '.js':'application/javascript','.css':'text/css','.json':'application/json',
  '.png':'image/png','.jpg':'image/jpeg','.svg':'image/svg+xml',
  '.ico':'image/x-icon','.woff2':'font/woff2','.woff':'font/woff',
  '.webp':'image/webp','.wasm':'application/wasm','.map':'application/json',
  '.ttf':'font/ttf',
};

function serveFile(fp, res, maxAge) {
  try {
    if (!fs.existsSync(fp)) return false;
    var s = fs.statSync(fp);
    if (!s.isFile()) return false;
    res.writeHead(200, {
      'Content-Type': MIME[path.extname(fp)] || 'application/octet-stream',
      'Content-Length': s.size,
      'Cache-Control': 'public, max-age=' + (maxAge != null ? maxAge : 31536000),
    });
    var data = fs.readFileSync(fp); res.end(data);
    return true;
  } catch (e) { return false; }
}

// Lazy Prisma - only loaded when truly needed
var prisma = null;
var prismaLoading = false;
function getPrisma() {
  if (prisma) return prisma;
  if (prismaLoading) return null;
  prismaLoading = true;
  try {
    var { PrismaClient } = require('@prisma/client');
    prisma = new PrismaClient({ log: ['error'] });
    console.log('[DB] Prisma loaded, RSS=' + Math.round(process.memoryUsage().rss/1024/1024) + 'MB');
    return prisma;
  } catch (e) {
    console.error('[DB] Prisma load failed:', e.message);
    return null;
  }
}

// Sample data for when Prisma isn't loaded yet
var sampleData = {
  users: [{ id: 'demo1', name: 'Miran M.', email: 'miran@rever.si', avatar: null, bike: 'Yamaha MT-07', bio: 'Motociklistični navdušenec' }],
  rides: [],
  routes: [],
  defaultUser: { id: 'demo1', name: 'Miran M.', email: 'miran@rever.si', avatar: null, bike: 'Yamaha MT-07', bio: 'Motociklistični navdušenec' },
  needsSeed: true,
  leaderboard: [{ id: 'demo1', name: 'Miran M.', totalDistance: 0, totalRides: 0 }],
};

var activeConns = 0;
var MAX_CONN = 10; // Process ONE request at a time!
var server = http.createServer(function(req, res) {
  activeConns++;
  if (activeConns > MAX_CONN) {
    activeConns--;
    res.writeHead(503, {'Content-Type': 'text/plain', 'Connection': 'close'});
    res.end('busy');
    return;
  }
  res.on('finish', function() { activeConns--; });
  res.on('close', function() { activeConns--; });
  
  console.log("[REQ] " + req.method + " " + req.url);
  try {
    var parsed = new URL(req.url, 'http://localhost');
    var pathname = parsed.pathname;
    var query = Object.fromEntries(parsed.searchParams);

    // 1. Homepage
    if (pathname === '/' || pathname === '') {
      res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8', 'Content-Length': cachedIndex.length, 'Cache-Control': 'public, max-age=60' });
      res.end(cachedIndex);
      return;
    }

    // 2. Static chunks
    if (pathname.startsWith('/_next/static/')) {
      var fp = path.join(__dirname, '.next', 'static', pathname.replace('/_next/static/', ''));
      if (serveFile(fp, res)) return;
      res.writeHead(404); res.end('Not Found');
      return;
    }

    // 3. Public assets
    if (pathname === '/sw.js' || pathname === '/manifest.json' || pathname === '/robots.txt') {
      if (serveFile(path.join(__dirname, 'public', pathname.slice(1)), res, 0)) return;
    }

    if (pathname.match(/\.(png|jpg|jpeg|svg|ico|woff2?|webp|json|txt|xml|mp3|wav)$/)) {
      if (serveFile(path.join(__dirname, 'public', pathname.slice(1)), res, 3600)) return;
    }

    // 4. API routes
    if (pathname.startsWith('/api/')) {
      try {
        var result = handleApi(req, res, pathname, query);
        if (result && !res.headersSent) {
          res.writeHead(result.code, result.headers);
          res.end(result.body);
        }
      } catch (e) {
        console.error('[API]', pathname, e.message);
        if (!res.headersSent) {
          res.writeHead(500, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ error: e.message }));
        }
      }
      return;
    }

    // 5. SPA fallback
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8', 'Content-Length': cachedIndex.length });
    res.end(cachedIndex);
  } catch (e) {
    console.error('[REQ]', e.message);
    if (!res.headersSent) { res.writeHead(500); res.end('Error'); }
  }
});

function json(code, data) {
  var body = JSON.stringify(data);
  return { code: code, headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) }, body: body };
}

function handleApi(req, res, pathname, query) {
  // Stub responses - no database, no async, instant responses
  if (pathname === "/api/init") {
    return json(200, { data: { users: [{ id: "demo1", name: "Miran M.", email: "miran@rever.si", avatar: null, bike: "Yamaha MT-07", bio: "Motociklistični navdušenec" }], rides: [], routes: [], defaultUser: { id: "demo1", name: "Miran M.", email: "miran@rever.si", avatar: null, bike: "Yamaha MT-07", bio: "Motociklistični navdušenec" }, needsSeed: true, leaderboard: [{ id: "demo1", name: "Miran M.", totalDistance: 0, totalRides: 0 }] } });
  }
  if (pathname === "/api/notifications") return json(200, { data: [] });
  if (pathname === "/api/weather") return json(200, { data: null });
  if (pathname === "/api/achievements") return json(200, { data: { earned: [], newlyEarned: [] } });
  if (pathname === "/api/sos") return json(200, { data: { ok: true } });
  if (pathname === "/api/seed") return json(200, { data: { seeded: false } });
  if (pathname === "/api/leaderboard") return json(200, { data: [{ id: "demo1", name: "Miran M.", totalDistance: 0, totalRides: 0 }] });
  if (pathname.startsWith("/api/users/")) return json(200, { data: { id: "demo1", name: "Miran M." } });
  if (pathname === "/api/rides") return json(200, { data: [] });
  if (pathname === "/api/routes") return json(200, { data: [] });
  if (pathname === "/api/comments") return json(200, { data: [] });
  if (pathname === "/api/settings") return json(200, { data: {} });
  // Default API response
  return json(200, { data: null });
}

function readBody(req) {
  return new Promise(function(resolve) {
    var b = '';
    req.on('data', function(c) { b += c; if (b.length > 1e6) req.destroy(); });
    req.on('end', function() { resolve(b); });
  });
}

server.maxConnections = 20;
server.timeout = 30000;
server.keepAliveTimeout = 65000;
server.headersTimeout = 66000;

server.listen(PORT, function() {
  console.log('> MotoTrack on :' + PORT + ' (RSS: ' + Math.round(process.memoryUsage().rss/1024/1024) + 'MB)');
});

// Memory monitor + GC
setInterval(function() {
  var m = process.memoryUsage();
  console.log('[MEM] RSS=' + Math.round(m.rss/1024/1024) + 'MB Heap=' + Math.round(m.heapUsed/1024/1024) + 'MB');
  if (global.gc && m.heapUsed > 80 * 1024 * 1024) {
    global.gc();
    console.log('[GC] Forced, new RSS=' + Math.round(process.memoryUsage().rss/1024/1024) + 'MB');
  }
}, 15000);
